Add docs/canonical-ways-to-investigate-bugs.md (#547)
LedgerHQ/ledger-live-common: docs/canonical-ways-to-investigate-bugs.md

If the bug is related to an UI issues or specific to the desktop platform, users can either contact our Customer support (preferred way) or [create an issue on GitHub](https://github.com/LedgerHQ/ledger-live-desktop).

If it's not, we generally will suspect the bugs can be reproduce in Live Common and using our CLI.

In that case, we will ask users to **Export Logs** in the app (either via a button or via Ctrl+E).

### It's a bug on Ledger Live Mobile?

If the bug is related to an UI issues, users can either contact our Customer support (preferred way) or [create an issue on GitHub](https://github.com/LedgerHQ/ledger-live-mobile).

We assume that most of the bugs on the mobile app, unless they are mobile platform specific or UI issues, are going to be common with what can be experienced on Desktop as we use the same stack shared with our Live Common library. We will therefore usually invite users to go back to desktop, try to reproduce the issue with the same set of accounts and issue or not we ask users to Export Logs from desktop (on our side we can try to export them back to mobile to see if we reproduce the issue).

## Investigate Ledger Live Desktop logs

> A log file from Ledger Live Desktop is essentially a list of JSON object separated by new lines.

Reviewer: ```suggestion
A logfile from Ledger Live Desktop is essentially a list of JSON objects separated by newlines.
```
Apparently "newline" is a noun

> From our [tools](../tools) project, we have a /logsviewer page that allow to visualize a log file:

Reviewer: ```suggestion
From our [tools](../tools) project, we have a /logsviewer page that allows visualizing a log file:
```

https://ledger-live-tools.now.sh/logsviewer

> This page allows to quickly browser and search through the logs.

Reviewer: ```suggestion
This page allows us to quickly browse and search through the logs.
```

> Logs are ordered from the most recent to the older.

Reviewer: ```suggestion
Logs are ordered from the most recent to the oldest.
```

> One of the important log that appears at the end is the "exportLogsMeta", which have some important user contextual information (Typically we want to check which envs has been used)

Reviewer: ```suggestion
One of the important log that appears at the end is the "exportLogsMeta", which has some important user contextual information (Typically we want to check which envs has been used)
```

<img width="1367" alt="Capture d’écran 2020-03-03 à 15 08 34" src="https://user-images.githubusercontent.com/211411/75783543-07a9c980-5d61-11ea-99d2-6c41e6076180.png">

Now depending on the problem being investigated you might want to look at the Ledger Live Desktop commands data

<img width="894" alt="Capture d’écran 2020-03-03 à 15 08 41" src="https://user-images.githubusercontent.com/211411/75783550-0b3d5080-5d61-11ea-861b-f7ffdcaf66b2.png">

or going more low level and look at the device binary exchanges

<img width="1294" alt="Capture d’écran 2020-03-03 à 15 09 18" src="https://user-images.githubusercontent.com/211411/75783551-0bd5e700-5d61-11ea-9c8f-2cb21234e03a.png">

> In any cases, our logs will also contain some analytics events (like the part of the app being navigated on) so we can have a big picture of what the user has been doing.

Reviewer: ```suggestion
In any case, our logs will also contain some analytics events (like the part of the app being navigated on) so we can have a big picture of what the user has been doing.
```

**On every investigation, we will usually want to reproduce on our side and it's often more practical to do it in the terminal with `ledger-live` CLI. Let's look at classical investigation examples.**

### A problem during account synchronisation

When a Synchronisation bug occurs, which usually is spotted by seeing a failure in the Sync icon or global notification, we can easily spot the problem in the logs.

> Usually, just seeing the logs will be enough, but sometimes we need to go deeper and try to reproduce on our side. We had a lot of Tezos bugs that were specific to some accounts, for instance, Tezos KT accounts with more than 100 transactions. It was easier to spot the problem by putting ourselves in the feet of our users.

Reviewer 2: *in the shoes of our users

> Usually, just seeing the logs will be enough, but sometimes we need to go deeper and try to reproduce in our side. For instance, we had a lot of Tezos bugs that was specific to some accounts, for instance Tezos KT accounts with more than 100 transactions. It was easier to spot the problem by putting ourself in the foot of our users.

Reviewer: ```suggestion
Usually, just seeing the logs will be enough, but sometimes we need to go deeper and try to reproduce on our side. We had a lot of Tezos bugs that were specific to some accounts, for instance, Tezos KT accounts with more than 100 transactions. It was easier to spot the problem by putting ourselves in the feet of our users.
```

**What you essentially will need is the Account xpub.** It can either be asked to the user or spotted in the logs:

<img width="976" alt="Capture d’écran 2020-03-03 à 16 01 07" src="https://user-images.githubusercontent.com/211411/75788726-ed73e980-5d68-11ea-9314-08f3468ab14b.png">

> > **It's important to know you will also need the [DerivationMode](derivation.md)**, in this specific case, it's a legacy account which is the default, but in any other case you will see the derivation mode after the xpub, for instance "segwit" or "native_segwit". You will have to add command parameter `-s segwit` (depending on the derivation mode)

Reviewer: ```suggestion
> **It's important to know you will also need the [DerivationMode](derivation.md)**, in this specific case, it's a legacy account which is the default, but in any other case you will see the derivation mode after the xpub, for instance, "segwit" or "native_segwit". You will have to add command parameter `-s segwit` (depending on the derivation mode)
```

> Now we the xpub, we can try to sync the account:

Reviewer 2: ```suggestion
Now that we have the xpub, we can try to sync the account:
```

{"message":"libcore-call: WalletPool#getWallet","data":["xpub6DBXGC1faXnWSWofjAXjfxihNmYEcJj7KcWytuc7AZK39zS8KdgbosCjnPyHRbwKxcJxnWbzLwZYaPHFE3zoQZmGDbPzwRuoSfFaeSTVwUT_zencash_"],"level":"debug","timestamp":"2020-03-03T15:12:36.063Z"}
{"message":"libcore-result: WalletPool#getWallet","data":{"value":{}},"level":"debug","timestamp":"2020-03-03T15:12:36.064Z"}
{"message":"libcore: getOrCreateAccount","data":{"xpub":"xpub6DBXGC1faXnWSWofjAXjfxihNmYEcJj7KcWytuc7AZK39zS8KdgbosCjnPyHRbwKxcJxnWbzLwZYaPHFE3zoQZmGDbPzwRuoSfFaeSTVwUT","index":0},"level":"debug","timestamp":"2020-03-03T15:12:36.065Z"}
{"message":"libcore-call: Wallet#getAccount","data":[0],"level":"debug","timestamp":"2020-03-03T15:12:36.065Z"}
{"message":"libcore-result: Wallet#getAccount","data":{"value":{}},"level":"debug","timestamp":"2020-03-03T15:12:36.133Z"}
{"message":"libcore: sync(1) syncCoreAccount","level":"debug","timestamp":"2020-03-03T15:12:36.133Z"}
{"message":"libcore-call: Account#synchronize","data":[],"level":"debug","timestamp":"2020-03-03T15:12:36.134Z"}
```

### A problem during "Add Account"

> The _Add Account_ flow involves the [`currencyBridge.scanAccounts`](CurrencyBridge.md) logic which itself is mostly about performing synchronisation. The main problem is that you might not know the XPUB at this point in time and it might breaks before. What we can do easily however is to reuse the APDUs that appears in the log and replay them against the CLI.

Reviewer 2: ```suggestion
The _Add Account_ flow involves the [`currencyBridge.scanAccounts`](CurrencyBridge.md) logic which itself is mostly about performing synchronisation. The main problem is that you might not know the XPUB at this point in time and it might break before. What we can do easily however is to reuse the APDUs that appears in the log and replay them against the CLI.
```

> On top of the /logsviewer page, there is a handy button "Export APDUs" that we can use to retrieve ALL the apdus. Now, you need to locate the part that concerns the add accounts.

Reviewer 2: No explanation of what APDUS are. Maybe we need to explain that they are the messages to communicate to the device ?

Author: good point! definitely need a apdu.md, i'll add a link for now

=> e04000000d038000002c8000007980000000
<= 41046a05e02c73991c84f42444cfa4b3bc1046838fe609a220b1498d766e23264d57ea4c4acaae65398c13adad5d045324e03abdeec9b3fef1d88d4abb832a44af9a237a6e53467933565a58506a48366b5944736b78544e65677a67365675634258317738466e39f205deb0a0a9763b86fecc61046a453be8a6763c8bdbceb8b3d8abff46459000
=> e04000000d038000002c8000007980000001
<= 4104392c448b6a725c7c8c32993d0483e38927830dfec299b74963d264b10e9db900215b1bf058a992604c86124f5d40c178b805da0f8af0889814cb7ffb011b4308237a6e62476d615656456d676d317a416e7a74724172617370506177574a73506a527a4af6c3ce6238696dc78ee8d57b161b25d79b89ace20515966789ac09f79c713cbd9000
```

We can save that to a `apdus.txt` file.

Now, we are going to our Terminal and we are going to replay that exact series of device exchanges. As everything is deterministic here, it should just works™.

We do this:

```
ledger-live proxy -f apdus.txt
```

which starts a server that will be able to respond like a device. It should logs something like

```
5 mocked APDUs will be replayed from /Users/grenaudeau/Desktop/logs.txt
DEVICE_PROXY_URL=ws://localhost:8435
DEVICE_PROXY_URL=ws://192.168.0.25:8435

Nano S proxy started on 192.168.0.25
```

> Now, we go to a new Terminal and we do the actual scan accounts via the `ledger-live sync -c ZEN` command. But we also need to make sure to use the environment DEVICE_PROXY_URL to use the proxy instead of a real device. (btw make sure to have removed `dbdata` folder that is used by ledger-live CLI)

Reviewer: ```suggestion
Now, we go to a new Terminal and we do the actual scan accounts via the `ledger-live sync -c ZEN` command. But we also need to make sure to use the environment `DEVICE_PROXY_URL` to use the proxy instead of a real device. (btw make sure to have removed `dbdata` folder that is used by ledger-live CLI)
```

```
DEVICE_PROXY_URL=ws://localhost:8435 ledger-live sync -c zen
```

> which correctly result of

Reviewer 2: ```suggestion
which correctly results in
```

      <= 410486f3033309829d1f1458622c6d08430f7130d4d1b2ddef378dfe4c1dd025dbbbeb51298dee5e7ad0c31a1f555fe8a2941bb0018185697276861622d0a3b1bcad237a6e6b346975357a65554a39616873554b677735635470446f374b32656554417a3473271f6a912e786b63d98809c92ba5c58beac445c698dcadb8c8e91652ce64745f9000
      => e04000000d038000002c8000007980000000
      <= 41046a05e02c73991c84f42444cfa4b3bc1046838fe609a220b1498d766e23264d57ea4c4acaae65398c13adad5d045324e03abdeec9b3fef1d88d4abb832a44af9a237a6e53467933565a58506a48366b5944736b78544e65677a67365675634258317738466e39f205deb0a0a9763b86fecc61046a453be8a6763c8bdbceb8b3d8abff46459000
      => e04000000d038000002c8000007980000001
      <= 4104392c448b6a725c7c8c32993d0483e38927830dfec299b74963d264b10e9db900215b1bf058a992604c86124f5d40c178b805da0f8af0889814cb7ffb011b4308237a6e62476d615656456d676d317a416e7a74724172617370506177574a73506a527a4af6c3ce6238696dc78ee8d57b161b25d79b89ace20515966789ac09f79c713cbd9000
      `
    }
  ]
};

export default dataset;
```

### A problem during Send flow, _before the device step_

Problems can appear at the beginning of the Send flow, typically when inputting the transaction data in the send form.

> We have in live-common a "TransactionStatus" concept which virtually everything is good to broadcast the transaction. The error is likely to be contained in that.

Reviewer 2: ```suggestion
We have in live-common a "TransactionStatus" concept which virtually checks everything is good to broadcast the transaction. The error is likely to be contained in that.
```

  transaction: '{"amount":"10000000000","recipient":"znaE1JH6YjjWiG23bqdCkBiC4KwZ9vo7hB6","useAllAmount":false,"family":"bitcoin","feePerByte":"1","networkInfo":{"family":"bitcoin","feeItems":{"items":[{"key":"0","speed":"high","feePerByte":"10"},{"key":"1","speed":"standard","feePerByte":"10"},{"key":"2","speed":"low","feePerByte":"10"}],"defaultFeePerByte":"10"}}}'
}
```

But if all is good, I will have some data:

```
{
  status: '{"errors":{},"warnings":{},"estimatedFees":"226","amount":"10000000","totalSpent":"10000226"}',
  transaction: '{"amount":"10000000","recipient":"znaE1JH6YjjWiG23bqdCkBiC4KwZ9vo7hB6","useAllAmount":false,"family":"bitcoin","feePerByte":"1","networkInfo":{"family":"bitcoin","feeItems":{"items":[{"key":"0","speed":"high","feePerByte":"10"},{"key":"1","speed":"standard","feePerByte":"10"},{"key":"2","speed":"low","feePerByte":"10"}],"defaultFeePerByte":"10"}}}'
}
```

> Just using `getTransactionStatus` will show you if a problem occur typically related to fetching fees or any other problem related to building a transaction. Now if the problem occurs at the device time, this will not be enough...

Reviewer 2: ```suggestion
Just using `getTransactionStatus` will show you if a problem occurs related to fetching fees or any other problem related to building a transaction. Now if the problem occurs at the device time, this will not be enough...
```

### A problem during send flow, _during the device step_

> Complex user's problem can appear during the Send flow. It often occur that the problem is specific to the user's Account and that we don't reproduce on our side.

Reviewer 2: ```suggestion
Complex user's problem can appear during the Send flow. It often occurs that the problem is specific to the user's Account and that we don't reproduce on our side.
```

In such case, the same technique can also be used by replaying APDUs. **Make sure to use `--disable-broadcast` to not broadcast the transaction.**

The technique involve both the sync and the replay logic. We need to know beforehand what the xpub because the transaction logic starts by having a synchronized account.

Essentially we can do this:

```
 DEVICE_PROXY_URL=ws://localhost:8435 ledger-live send --disable-broadcast -c ZEN --xpub xpub6DBXGC1faXnWSWofjAXjfxihNmYEcJj7KcWytuc7AZK39zS8KdgbosCjnPyHRbwKxcJxnWbzLwZYaPHFE3zoQZmGDbPzwRuoSfFaeSTVwUT --recipient znaE1JH6YjjWiG23bqdCkBiC4KwZ9vo7hB6 --amount 0.1
```

> By using the exact same parameter as the user we could be able to reproduce, even tho in practice we will likely fall into problems because transactions tends to not be identically reproductible / deterministic.

Reviewer 2: ```suggestion
By using the exact same parameter as the user we could be able to reproduce, even though in practice we will likely fall into problems because transactions tends to not be identically reproductible / deterministic.
```
